source/common/http/filter/cors_filter.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "envoy/http/filter.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * Configuration for the cors filter.
+ */
+struct CorsFilterConfig {};
+
+typedef std::shared_ptr<const CorsFilterConfig> CorsFilterConfigConstSharedPtr;
+
+class CorsFilter : public StreamFilter, Logger::Loggable<Logger::Id::filter> {
+public:
+  CorsFilter(CorsFilterConfigConstSharedPtr config);
+  ~CorsFilter();
+
+  void initialize();
+
+  // Http::StreamFilterBase
+  void onDestroy() override{};
+
+  // Http::StreamDecoderFilter
+  FilterHeadersStatus decodeHeaders(HeaderMap& headers, bool end_stream) override;
+  FilterDataStatus decodeData(Buffer::Instance&, bool) override {
+    return FilterDataStatus::Continue;
+  };
+  FilterTrailersStatus decodeTrailers(HeaderMap&) override {
+    return FilterTrailersStatus::Continue;
+  };
+  void setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) override;
+
+  // Http::StreamEncoderFilter
+  Http::FilterHeadersStatus encodeHeaders(Http::HeaderMap& headers, bool end_stream) override;
+  Http::FilterDataStatus encodeData(Buffer::Instance&, bool) override {
+    return FilterDataStatus::Continue;
+  };
+  Http::FilterTrailersStatus encodeTrailers(Http::HeaderMap&) override {
+    return FilterTrailersStatus::Continue;
+  };
+  void setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callbacks) override {
+    encoder_callbacks_ = &callbacks;
+  };
+
+private:
+  friend class CorsFilterTest;
+
+  const std::string& allowOrigin();
+  const std::string& allowMethods();
+  const std::string& allowHeaders();
+  const std::string& exposeHeaders();
+  const std::string& maxAge();
+  bool allowCredentials();
+  bool enabled();
+
+  CorsFilterConfigConstSharedPtr config_;
+  StreamDecoderFilterCallbacks* decoder_callbacks_{};
+  StreamEncoderFilterCallbacks* encoder_callbacks_{};
+  const Envoy::Router::CorsPolicy* routeCorsPolicy_{};
```

Comment:
member variables get snake case so route_cors_policy_{} and virtual_host_cors_policy_{}

Meanwhile Envoy isn't big on raw pointers where avoidable* and I'm unsure this would be safe if there were a config reload.  Should they be shared pointers instead?

*there's totally precedent for the Encoder/Decoder callbacks being raw pointers so don't worry about those.

Reply:
I have a shared pointer version stashed. Let me apply and push it. I hesitated adding it because no other policy / config used it. I added it as shared pointers all the way from config_impl router::routeentry and router::virtualhost. Not sure that's all correct.